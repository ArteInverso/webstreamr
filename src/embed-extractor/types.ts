import { StreamWithMeta } from '../types';

export interface EmbedExtractor {
  readonly id: string;

  readonly label: string;

  readonly extract: (embedUrl: string, language: string) => Promise<StreamWithMeta>;
}
