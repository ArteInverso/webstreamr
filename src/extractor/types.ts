import { Context, Meta, UrlResult } from '../types';

export interface Extractor {
  readonly id: string;

  readonly label: string;

  readonly ttl: number;

  readonly supports: (ctx: Context, url: URL) => boolean;

  readonly extract: (ctx: Context, url: URL, meta: Meta) => Promise<UrlResult[]>;
}
