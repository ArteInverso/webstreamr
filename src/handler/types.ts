import { Context, UrlResult } from '../types';

export interface Handler {
  readonly id: string;

  readonly label: string;

  readonly contentTypes: string[];

  readonly languages: string[];

  readonly handle: (ctx: Context, type: string, id: string) => Promise<(UrlResult | undefined)[]>;
}
