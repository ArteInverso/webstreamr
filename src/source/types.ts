import { ContentType } from 'stremio-addon-sdk';
import { Context, CountryCode } from '../types';
import { Id } from '../utils';

export interface SourceResult {
  countryCode: CountryCode;
  title?: string;
  url: URL;
}

export interface Source {
  readonly id: string;

  readonly label: string;

  readonly contentTypes: ContentType[];

  readonly countryCodes: CountryCode[];

  handle(ctx: Context, type: ContentType, id: Id): Promise<(SourceResult[])>;
}
