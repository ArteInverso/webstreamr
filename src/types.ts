import { Manifest, ManifestConfig } from 'stremio-addon-sdk';

export interface Context {
  id: string;
  ip: string;
  config: Config;
}

export type ManifestWithConfig = Manifest & { config: ManifestConfig[] };

export type Config = Record<string, string>;

export interface Meta {
  bytes?: number;
  countryCode: string;
  height?: number;
}

export interface UrlResult {
  url: URL;
  isExternal?: boolean;
  label: string;
  sourceId: string;
  meta: Meta;
  requestHeaders?: Record<string, string>;
}
