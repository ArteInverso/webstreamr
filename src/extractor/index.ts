import { DoodStream } from './DoodStream';
import { Dropload } from './Dropload';
import { ExternalUrl } from './ExternalUrl';
import { Extractor } from './Extractor';
import { Fetcher } from '../utils';
import { KinoGer } from './KinoGer';
import { Mixdrop } from './Mixdrop';
import { Soaper } from './Soaper';
import { SuperVideo } from './SuperVideo';
import { Uqload } from './Uqload';
import { VidSrc } from './VidSrc';
import { VixSrc } from './VixSrc';

export * from './Extractor';
export * from './ExtractorRegistry';

export const createExtractors = (fetcher: Fetcher): Extractor[] => [
  new DoodStream(fetcher),
  new Dropload(fetcher),
  new SuperVideo(fetcher),
  new KinoGer(fetcher),
  new Mixdrop(fetcher),
  new Soaper(fetcher),
  new Uqload(fetcher),
  new VidSrc(fetcher),
  new VixSrc(fetcher),
  new ExternalUrl(fetcher), // fallback extractor which must come last
];
