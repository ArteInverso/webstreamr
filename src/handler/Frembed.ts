import { ContentType } from 'stremio-addon-sdk';
import { Handler, HandleResult } from './types';
import { Fetcher, getTmdbId, Id } from '../utils';
import { Context, CountryCode } from '../types';

export class Frembed implements Handler {
  readonly id = 'frembed';

  readonly label = 'Frembed';

  readonly contentTypes: ContentType[] = ['series'];

  readonly countryCodes: CountryCode[] = [CountryCode.fr];

  private readonly fetcher: Fetcher;

  constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  readonly handle = async (ctx: Context, _type: string, id: Id): Promise<HandleResult[]> => {
    const tmdbId = await getTmdbId(ctx, this.fetcher, id);

    const apiUrl = new URL(`https://frembed.space/api/series?id=${tmdbId.id}&sa=${tmdbId.season}&epi=${tmdbId.episode}&idType=tmdb`);

    const json = JSON.parse(await this.fetcher.text(ctx, apiUrl));

    const urls: URL[] = [];
    for (const key in json) {
      if (key.startsWith('link') && json[key] && !json[key].includes(',https')) {
        try {
          urls.push(new URL(json[key].trim()));
        } catch {
          // Skip invalid URL
        }
      }
    }

    return urls.map(url => ({ countryCode: CountryCode.fr, referer: apiUrl, title: `${json['title']} ${tmdbId.season}x${tmdbId.episode}`, url }));
  };
}
