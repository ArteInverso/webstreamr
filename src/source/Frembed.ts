import { ContentType } from 'stremio-addon-sdk';
import { Context, CountryCode } from '../types';
import { Fetcher, getTmdbId, Id } from '../utils';
import { Source, SourceResult } from './types';

export class Frembed implements Source {
  public readonly id = 'frembed';

  public readonly label = 'Frembed';

  public readonly contentTypes: ContentType[] = ['series'];

  public readonly countryCodes: CountryCode[] = [CountryCode.fr];

  private readonly fetcher: Fetcher;

  public constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  public async handle(ctx: Context, _type: string, id: Id): Promise<SourceResult[]> {
    const tmdbId = await getTmdbId(ctx, this.fetcher, id);

    const apiUrl = new URL(`https://frembed.wiki/api/series?id=${tmdbId.id}&sa=${tmdbId.season}&epi=${tmdbId.episode}&idType=tmdb`);

    const json = JSON.parse(await this.fetcher.text(ctx, apiUrl));

    const urls: URL[] = [];
    for (const key in json) {
      if (key.startsWith('link') && json[key] && !json[key].includes(',https')) {
        try {
          urls.push(new URL(json[key].trim()));
        } catch {
          // Skip invalid URL
        }
      }
    }

    return urls.map(url => ({ countryCode: CountryCode.fr, title: `${json['title']} ${tmdbId.season}x${tmdbId.episode}`, url }));
  };
}
