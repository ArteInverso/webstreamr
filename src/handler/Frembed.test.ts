import winston from 'winston';
import { Frembed } from './Frembed';
import { ExtractorRegistry } from '../extractor';
import { Fetcher } from '../utils';
import { Context } from '../types';
jest.mock('../utils/Fetcher');

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
// @ts-expect-error No constructor args needed
const fetcher = new Fetcher();
const handler = new Frembed(fetcher, new ExtractorRegistry(logger, fetcher));
const ctx: Context = { id: 'id', ip: '127.0.0.1', config: { fr: 'on' } };

describe('Frembed', () => {
  test('does not handle non imdb series', async () => {
    const streams = await handler.handle(ctx, 'series', 'kitsu:123');

    expect(streams).toHaveLength(0);
  });

  test('handles non-existent series gracefully', async () => {
    const streams = await handler.handle(ctx, 'series', 'tt4352342:1:1');

    expect(streams).toHaveLength(0);
  });

  test('handle imdb black mirror s4e2', async () => {
    const streams = (await handler.handle(ctx, 'series', 'tt2085059:4:2')).filter(stream => stream !== undefined);

    expect(streams).toHaveLength(1);
    expect(streams[0]).toStrictEqual({
      url: expect.any(URL),
      label: 'DoodStream',
      sourceId: 'doodstream_fr',
      height: 0,
      bytes: 0,
      countryCode: 'fr',
      requestHeaders: {
        Referer: 'http://dood.to/',
      },
    });
    expect(streams[0]?.url.href).toMatch(/^https:\/\/.*?token.*?expiry/);
  });

  test('handle imdb black mirror s7e1', async () => {
    const streams = (await handler.handle(ctx, 'series', 'tt2085059:7:1')).filter(stream => stream !== undefined);

    expect(streams).toHaveLength(1);
    expect(streams[0]).toStrictEqual({
      url: expect.any(URL),
      label: 'UQLoad',
      sourceId: 'uqload_fr',
      height: 0,
      bytes: 0,
      countryCode: 'fr',
      requestHeaders: {
        Referer: 'https://uqload.net',
      },
    });
    expect(streams[0]?.url.href).toMatch(/^https:\/\/.*?uqload\.net\/.*?\.mp4/);
  });
});
