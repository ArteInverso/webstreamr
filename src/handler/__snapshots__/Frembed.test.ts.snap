// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Frembed handle imdb black mirror s4e2 1`] = `
[
  {
    "bytes": 0,
    "countryCode": "fr",
    "height": 0,
    "label": "DoodStream",
    "requestHeaders": {
      "Referer": "http://dood.to/",
    },
    "sourceId": "doodstream_fr",
    "url": "https://ee317r.cloudatacdn.com/u5kj63yvxddlsdgge7qremqqka27irwiupc4k7o5cikbh7fdq4j4mwzraf7q/1y4sn4ndoi~undefined?token=fh3rdlhjvao7chgxndsi3ra7&expiry=639837296000",
  },
  {
    "bytes": 0,
    "countryCode": "fr",
    "height": 0,
    "isExternal": true,
    "label": "netu.frembed.art",
    "sourceId": "external_fr",
    "url": "https://netu.frembed.art/e/0DFgfkcXOsDP",
  },
  {
    "bytes": 0,
    "countryCode": "fr",
    "height": 0,
    "isExternal": true,
    "label": "johnalwayssame.com",
    "sourceId": "external_fr",
    "url": "https://johnalwayssame.com/e/cqy9oue7sv0g",
  },
]
`;
